### ISSUE TITLE
Unit inconsistency with QMMM charges

### ISSUE THREAD

**Issue by tovrstra** [psi4/psi4#1881]
I would expect the two attached calculations to give the same result but they don't. The difference is that in one input, the geometry is in angstroms, while in the other the same geometry is specified in Bohr. Somehow, the term `Additional nuclear repulsion` is computed very differently:

```
$ grep "nuclear" */psi4.out
case1/psi4.out:  Old nuclear repulsion        =    0.529177210670000
case1/psi4.out:  Additional nuclear repulsion =    0.330403525088017
case1/psi4.out:  Total nuclear repulsion      =    0.859580735758017
case2/psi4.out:  Old nuclear repulsion        =    0.529177525830478
case2/psi4.out:  Additional nuclear repulsion =    0.793766288745717
case2/psi4.out:  Total nuclear repulsion      =    1.322943814576195
```

In- and outputs attached: [cases.zip](https://github.com/psi4/psi4/files/4586584/cases.zip)

Input files for convenience:

`case1/psi4.in`

```
molecule mol {
0 1
H         0.0   0.0   0.0
H         0.0   0.0   1.0
units angstrom
symmetry c1
no_reorient
no_com
}

set basis 3-21G

chfield = QMMM()
chfield.extern.addCharge(1.0, 0.0, 0.0, 3.77945)
psi4.set_global_option_python('EXTERN', chfield.extern)

energy("SCF")
```

`case2/sp4i.in`

```
molecule mol {
0 1
H         0.0   0.0   0.0
H         0.0   0.0   1.0
units angstrom
symmetry c1
no_reorient
no_com
}

set basis 3-21G

chfield = QMMM()
chfield.extern.addCharge(1.0, 0.0, 0.0, 3.77945)
psi4.set_global_option_python('EXTERN', chfield.extern)

energy("SCF")
```

The results for `case2` seem correct while `case1` seems wrong. In both cases, the output mentions the same position of the external charge in atomic untis:

`case1/psi4.out`

```
...
   => External Potential Field:  <= 

    > Charges [a.u.] < 

              Z          x          y          z
        1.00000    0.00000    0.00000    3.77945
...
```

`case2/psi4.out`

```
...
   => External Potential Field:  <= 

    > Charges [a.u.] < 

              Z          x          y          z
        1.00000    0.00000    0.00000    3.77945
...
```

**Comment by hokru** [comment]
Could you edit your post to have the right inputs? Right now both cases specify `angstrom` (your zipped inputs are different).

> The atomic charge is specified in atomic units, and the coordinates always use the same units as the geometry specification in the regular QM region

Which means your charge sits at different positions in case1/2 since in both you use `x-coordinate = 1.0`. Unless I misunderstand the problem at hand.

**Comment by tovrstra** [comment]
@hokru Apologies for the copy-paste mistake. I fixed it.

Could it be that the unit in the molecule block also affects the unit of the external charge coordinates in the input and in the output (despite the fact that the output mentions a.u. in both cases)? That could explain the differences.

**Comment by loriab** [comment]
I had to change this within the past few months. Let me see when and where, as it may have been only in the imminent DDD branch.

**Comment by hokru** [comment]
No worries. I quoted the manual above. The external charge unit should change according to the coordinates in the input yes. 

The output seems inconsistent to the manual statement.

**Comment by loriab** [comment]
Ok, this was changed in Nov 2019 in #1351 (unmerged) to "extern object units are bohr, not whatever the Molecule was." And in that branch Mol units ang/bohr + extern units bohr are behaving sensibly:

```
molecule mol {
0 1
H         0.0   0.0   0.0
H         0.0   0.0   1.0
units angstrom
symmetry c1
no_reorient
no_com
}

set basis 3-21G

embedding_charges = [
    [1.0, [0.0, 0.0, 3.77945]],
]

energy("SCF", embedding_charges=embedding_charges)
```


```
>>> diff toon1.out toon2.out 
40,41c40,41
< H         0.0   0.0   1.0
< units angstrom
---
> H         0.0   0.0   1.889725
> units bohr
85c85
<     Geometry (in Angstrom), charge = 0, multiplicity = 1:
---
>     Geometry (in Bohr), charge = 0, multiplicity = 1:
90c90
<          H            0.000000000000     0.000000000000     1.000000000000     1.007825032230
---
>          H            0.000000000000     0.000000000000     1.889725000000     1.007825032230
94,96c94,96
<   Rotational constants: A = ************  B =     16.72674  C =     16.72674 [cm^-1]
<   Rotational constants: A = ************  B = 501455.10648  C = 501455.10648 [MHz]
<   Nuclear repulsion =    0.529177210670000
---
>   Rotational constants: A = ************  B =     16.72676  C =     16.72676 [cm^-1]
>   Rotational constants: A = ************  B = 501455.70378  C = 501455.70378 [MHz]
>   Nuclear repulsion =    0.529177525830478
165,167c165,167
<   Old nuclear repulsion        =    0.529177210670000
<   Additional nuclear repulsion =    0.793766603906570
<   Total nuclear repulsion      =    1.322943814576570
---
>   Old nuclear repulsion        =    0.529177525830478
>   Additional nuclear repulsion =    0.793766288745717
>   Total nuclear repulsion      =    1.322943814576195 
215c215
<   @DF-RHF Final Energy:    -1.12042760680024
---
>   @DF-RHF Final Energy:    -1.12042770344228
219,222c219,222
<     Nuclear Repulsion Energy =              1.3229438145765702
<     One-Electron Energy =                  -3.0366720199640036
<     Two-Electron Energy =                   0.5933005985871914
<     Total Energy =                         -1.1204276068002419
---
>     Nuclear Repulsion Energy =              1.3229438145761949
>     One-Electron Energy =                  -3.0366721911846972
>     Two-Electron Energy =                   0.5933006731662177
>     Total Energy =                         -1.1204277034422847
```

Bad news is current master does indeed seem wrong. Since the DDD fix is waiting, can you work with your case 2 setup in master for now?

**Comment by tovrstra** [comment]
Thanks for all the comments. My confusion is solved, which is definitely a good pragmatic solution since this was just a test for prototyping. It is not an urgent issue and if anyone else runs into this, it is easy to find the workaround here.

**Comment by fevangelista** [comment]
Just to clarify, is the long-term plan to use Bohr for external charges irrespective of the molecule units? I'd rather have both specified in the same units to avoid confusion and to make input creation easy.

**Comment by loriab** [comment]
@fevangelista correct, future plan is to use Bohr. An ang input example is https://github.com/psi4/psi4/pull/1351/files#diff-87670b13951f7da58a7e608cbf6309a1 . I did try to keep the current tie-units-to-Mol pattern, but it didn't work easily. And with the charges passed in as argument, it seemed disconnected from Mol. Too objectionable?

**Comment by tovrstra** [comment]
I'm going a bit of topic. I tried to understand how the unit conventions were handled in PSI4 and I'd like to share the things I ran into.

I'd suggest to keep it simple and use atomic units everywhere by default. It would be very helpful when any deviation from such a basic rule is made explicit. To keep it simple, the usage of non-atomic units can further limited to reading inputs or writing output files. At the moment, this does not seem to be the case, which is confusing.

Some examples:

- The point that confused me in this issue: changing the unit to Angstrom in the molecule section also affects units in the QMMM charges. The latter is an implicit unit convention. I'm glad this is being fixed. If you want to support Angstrom from QMMM charges, add an explicit option for it. Something along the following lines could already be a simple solution:
   ```python
   embedding_charges = [
       [1.0, array([0.0, 0.0, 2.0])*angstrom],
   ]
   ```

- In `case1/psi4.out`, the positions of the QMMM charges are printed in Angstrom while the preceding header mentions a.u. Also here, the unit convention is implicit, easily leading to confusion.

- When looking e.g. at libmints, there seem to be unit conversions buried deep into the code, e.g. related to this issue: https://github.com/psi4/psi4/blob/bf127d90c4b40f2ab6471bd6de4f312c9a149b1d/psi4/src/psi4/libmints/extern.cc#L306-L323
  That is just asking for trouble. The only good reason not to use atomic units is to interface with human beings or certain data containers following other conventions. I would suggest to avoid dragging unit conventions all the way into computational parts of the code.

**Comment by fevangelista** [comment]
@loriab and @tovrstra: Can we find a compromise where we keep the interface consistent but make sure that the inner workings of C++ classes is all in a.u.? I agree that have the logic deep down the code is not the right way to go.

I would keep the current automatic conversion for several reasons:
1. **Continuity**. People are already using it, so switching could create some challenges since one would have to be careful when setting up computations for different version of psi4
1. **Potential confusion**. The potential for confusing bohr/ang with the external field is greater than that of confusing bohr/ang in the molecule geometry. For molecules if one confuses bohr vs. ang you easily run into convergence problems or get nonsensical results. But for the external field corrections it's not obvious, unless one does some testing. We, for example, could entirely miss this change in our computations.
1. **Practicality**. We use the external charges and in our typical workflow we prepare a geometry for a lattice of atoms and then separate the QM from the MM region. It's just simpler to keep everything consistent in one set of units.

I have to say, that from the user perspective, the current way we handle the input is not ideal. I am not familiar with QMMM codes, but what would be the most convenient input format is something like we have for the molecular fragments:
```
molecule {
C 
O 
<some delimiter>
Q X1 Y1 Z1 CHARGE1
Q X2 Y2 Z2 CHARGE2
}
```

**Comment by hokru** [comment]
For QM/MM, since you have many thousands of points, most QM programs use a simple external file for point charges so one does not have to send everything through the slower-but-smarter "molecule recognition" machinery. Besides, these external files are easier to write by the MM program interface as well.

**Comment by tovrstra** [comment]
@fevangelista I agree with all your points and don't have much to add. In case of external files, is there something well-established?

**Comment by loriab** [comment]
Because of the switch from `QMMM()` obj to `embedded_charges` arg (req'd for DDD), all current inputs will break anyway. Upgrade guidance is already in place: https://github.com/psi4/psi4/pull/1351/commits/232d0378f893d619ecf16d95fc59281e948fbb23#diff-6c69f4c34c8ea7d7015fd7abef2e5a88R123 . Since reason can't guide one to a sensible default, I'm against the units of Mol influencing anything outside the `molecule {...}` block. I agree that the `Q X1 Y1 Z1 CHARGE1` is appealing, but I think we want to address this as a whole with all the multipoles. Andy, Daniel, and I got started on this for a day last September, then L2 took over. Schema started over at https://github.com/MolSSI/QCElemental/pull/133 .

**Comment by hokru** [comment]
You mean a standard format for point charge files for QMMM? Not that I know of. 
I know the AMBER QM/MM interface a bit, and there are all kinds of formats for different QM programs.

**Comment by fevangelista** [comment]
Ok, I still think the solution is a bit inconsistent. Shouldn't we also take Mol coordinates in Bohr then? or that we use Angstrom in the embedded_charges for consistency with the Mol default? Of course Bohr to And conversion is a problem by itself.
